### ISSUE TITLE
Bypass using execCommand

### ISSUE THREAD

**Issue by apple502j** [w3c/webappsec-trusted-types#345]
Trusted Types can be bypassed using `document.execCommand` with `insertHTML` as the first parameter and the injected content as the third argument. While `document.execCommand` does not have an official specification and is deprecated, "[u]ser agents are encouraged to implement the features described in execCommand"([WHATWG HTML spec](https://html.spec.whatwg.org/multipage/infrastructure.html#dependencies)) and for webcompat reasons I don't expect this function or insertHTML command to be removed. IMO execCommand should be restricted under Trusted Types - execution of insertHTML requires TrustedHTML as the third parameter.

PoC: [trusted-types-execcommand.zip](https://github.com/w3c/webappsec-trusted-types/files/6839199/trusted-types-execcommand.zip)

**Comment by koto** [comment]
Thanks so much!

I think we should indeed guard this. Not sure if this can be done normatively, but we should mention that issue in the spec, and follow with the implementation. Given that this is a new sink, we'll investigate how to add it without breaking the websites, perhaps make the CSP report-only the violation for a couple of browser releases.

Are other commands also affected? E.g. `paste`?

**Comment by koto** [comment]
I filed https://bugs.chromium.org/p/chromium/issues/detail?id=1230567 for the Chromium implementation bits.

**Comment by apple502j** [comment]
I don't think we can use `paste` outside of extensions with `clipboardRead` permission. There's also a built-in sanitizer which should prevent the "worst case" (DOM XSS).

`createLink` accepts `javascript:` URL but that's already blocked by Trusted Types at least on Chromium.

`defaultParagraphSeparator` accepts tag name but I don't think it accepts `script` on Firefox and is not implemented anyway on Chromium. `formatBlock` exists on Chromium too but I can't use `script` there.

**Comment by koto** [comment]
Thanks!

`formatBlock` should be OK, there's an [allowlist for tag names](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/editing/commands/format_block_command.cc;l=174;drc=23ee32391a2364af6e4ed7573f4b2d2e380e441b). We also reviewed and `insertHTML` seems to be the only one relevant.
